### ISSUE TITLE
Type-stability in Chebyshev polynomials

### ISSUE THREAD

**Issue by jishnub** [jverzani/SpecialPolynomials.jl#44]
Since the degree and variable are stored as type parameters, this makes the constructor inherently type-unstable.

```julia
julia> @code_warntype Chebyshev([1])
MethodInstance for Chebyshev(::Vector{Int64})
  from Chebyshev(coeffs::Vector{var"#172#S"}) where var"#172#S" in SpecialPolynomials at /home/jishnu/Dropbox/PostDoc/RossbyWaves/RossbyWaveSpectrum.jl/test/dev/SpecialPolynomials/src/Orthogonal/abstract.jl:50
Static Parameters
  #172#S = Int64
Arguments
  #self#::Type{Chebyshev}
  coeffs::Vector{Int64}
Body::Chebyshev{Int64}
1 ─ %1 = (#self#)(coeffs, :x)::Chebyshev{Int64}
└──      return %1
```

I haven't looked into the other polynomial types, but I suspect similar issues would exist elsewhere too.

**Comment by jverzani** [comment]
Yeah, this was a design choice. Maybe not the best. The idea was by fixing the degree in the type, we could generate faster function evaluations and other operations using generated functions. This worked really well with the `ImmutablePolynomial` class in Polynomials, but also means those polynomials are poor choices for making arrays of polynomials. But let me look at your PR to see what you have suggested.

**Comment by jverzani** [comment]
This should be (mostly) addressed in #48. Please re-open if you find it needs more. Thanks!
